Allow config to be provided via command-line

cameronhunter opened the pull request:
### Summary
I use `lint-staged` as a pre-commit hook and would like to explicitly provide the config file location. This PR uses [`commander`](npmjs.com/package/commander) to parse the command-line argument and passes it as `cosmiconfig`'s [`configPath`](https://github.com/davidtheclark/cosmiconfig#configpath) option. This allows the config file to be any format already supported.

### Test plan
I've added tests and updated the documentation.

sudo-suhas commented:
@cameronhunter Could you investigate the CI failures?

cameronhunter commented:
@sudo-suhas The CI failures were:

1. Cross-platform difference in paths between Windows and Linux. Resolved using a jest serializer.
2. In some CI setups `Listr` decides to be verbose and write more to the console, which meant the snapshots were inconsistent. Resolved by injecting the console logger instead of mocking the global console.

sudo-suhas commented:
Hey @cameronhunter, thanks for this. I did not know about the jest snapshot serializer. Cool stuff.

Regarding 99d44b9, do you think it is possible that the whitespace issue was because the editor removed trailing whitespace? I personally have that enabled in vscode.

> In some CI setups `Listr` decides to be verbose and write more to the console

Actually, I don't think this is the case. I am only guessing, but I think this has something to do with running the code in sync vs async. In all our current tests, since `cosmiconfig` was being mocked, we only captured the console logs which were printed by us. And _after_ that `listr` wrote some more logs. However, in the test you added, cosmiconfig ran async and by the time we `await`ed for `lintStaged`, `listr` had already pushed some logs to console. For some reason, this isn't happening locally. Anyway, `listr` behavior isn't different but our way of capturing `console` logs is buggy. So I think it is the right approach to inject the logger. However, I think a lot of complexity was added in the way it has been implemented now. I took some of your commits and tweaked them in #311. Could you please remove commits ce2c8d7, 327f1c7, 58b034a, 505654a, 40b399a, afe2fac and port the changes from d512c63 instead? You can use [git reset](https://stackoverflow.com/questions/1338728/delete-commits-from-a-branch-in-git) to remove commits. Let me know if you need any help.

cameronhunter commented:
@sudo-suhas I've rebased on `master`, squashed to a single commit and used `npm run cz` for formatting the commit message. 

I reordered the parameters to the main lib to pass the logger first (with a default value of `console`) so that it doesn't cause a breaking change. Overall it this should just be a minor version bump.

CI is passing – let me know if there's anything more you'd like included before merging.

sudo-suhas commented on test/__mocks__/my-config.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "verbose": false,
```
Could you change this to `true`? We'd be able to see the config in the snapshot then.

cameronhunter replied:
Sure

cameronhunter commented:
@sudo-suhas Can you speak to a timeline of merging and publishing the new version?

sudo-suhas commented:
I figured I'd wait a couple of days for comments, if any, from @okonet and @luftywiranda13. I'll wait till end of day and merge it in. It should be released immediately via semantic-release.

luftywiranda13 commented:
is it better if we also add some additional helpful things for the `--help` page?

sudo-suhas commented:
@luftywiranda13 I think that makes sense.

sudo-suhas commented:
@cameronhunter This has been [released in `v4.3.0`](https://github.com/okonet/lint-staged/releases/tag/v4.3.0). Apologies for the delay and thank you for your contribution!
